### PR TITLE
fixed bug for injection analysis if Istio is installed with sidecarInjectorWebhook.enableNamespacesByDefault=true

### PR DESCRIPTION
**Pull request by zhlsunshine**
According to [doc](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#controlling-the-injection-policy), namespaces or pods should be treated as injectable if Istio is installed with sidecarInjectorWebhook.enableNamespacesByDefault=true and there is no any injected label is set.

This bug fixed 2 issues:

1. removed unnecessary information for injection analysis for some systems namespaces, such as istio-system, kube-system etc.
2. check the injection option `sidecarInjectorWebhook.enableNamespacesByDefault` if no label is set.
There is also new unit test for `sidecarInjectorWebhook.enableNamespacesByDefault` case.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [x] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by zhlsunshine** [comment]
Here are some test results:

**sidecarInjectorWebhook.enableNamespacesByDefault is false**:

```
$ istioctl install --set profile=demo -y
$ istioctl analyze -A
......
Info [IST0102] (Namespace bookinfo) The namespace is not enabled for Istio injection. Run 'kubectl label namespace bookinfo istio-injection=enabled' to enable it, or 'kubectl label namespace bookinfo istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service calico-kube-controllers-metrics.calico-system) Port name metrics-port (port: 9094, targetPort: 9094) doesn't follow the naming convention of Istio port.
......
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.12/docs/reference/config/analysis for more information about causes and resolutions.
```

**sidecarInjectorWebhook.enableNamespacesByDefault is true**:

```
$ istioctl install --set profile=demo -y --set values.sidecarInjectorWebhook.enableNamespacesByDefault=true
$ istioctl analyze -A
......
Info [IST0118] (Service calico-kube-controllers-metrics.calico-system) Port name metrics-port (port: 9094, targetPort: 9094) doesn't follow the naming convention of Istio port.
......
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.12/docs/reference/config/analysis for more information about causes and resolutions.
```

File: galley/pkg/config/analysis/analyzers/injection/injection.go [hunk 1]
```diff
@@ -51,16 +52,28 @@ func (a *Analyzer) Metadata() analysis.Metadata {
...
 // Analyze implements Analyzer
 func (a *Analyzer) Analyze(c analysis.Context) {
+	enableNamespacesByDefault := false
 	injectedNamespaces := make(map[string]bool)
 
+	// verify the enableNamespacesByDefault flag in injection configmaps
+	c.ForEach(collections.K8SCoreV1Configmaps.Name(), func(r *resource.Instance) bool {
+		if r.Metadata.FullName.Name.String() == util.InjectionConfigMap {
+			cm := r.Message.(*v1.ConfigMap)
+			enableNamespacesByDefault = GetEnableNamespacesByDefaultFromInjectedConfigMap(cm)
```
Comment (irisdingbj) [inline galley/pkg/config/analysis/analyzers/injection/injection.go:69]: need to consider revision info when getting the cm name.

File: galley/pkg/config/analysis/analyzers/injection/injection.go [hunk 2]
```diff
@@ -73,8 +86,11 @@ func (a *Analyzer) Analyze(c analysis.Context) {
 		_, okNewInjectionLabel := r.Metadata.Labels[RevisionInjectionLabelName]
 
 		if injectionLabel == "" && !okNewInjectionLabel {
-			// TODO: if Istio is installed with sidecarInjectorWebhook.enableNamespacesByDefault=true
+			// if Istio is installed with sidecarInjectorWebhook.enableNamespacesByDefault=true
 			// (in the istio-sidecar-injector configmap), we need to reverse this logic and treat this as an injected namespace
+			if enableNamespacesByDefault {
+				return true
```
Comment (irisdingbj) [inline galley/pkg/config/analysis/analyzers/injection/injection.go:92]: we might need to tell user injecition is enabled by this flag on all namespaces.

File: galley/pkg/config/analysis/analyzers/injection/injection.go (hunk 1 above)
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection.go:69]: done, move this section to namespace loop to retrieve specified revision.

File: galley/pkg/config/analysis/analyzers/injection/injection.go (hunk 2 above)
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/injection/injection.go:92]: done, added a new message type named `NamespaceInjectionEnabledByDefault` to report this information to user.

**Comment by zhlsunshine** [comment]
Hi @irisdingbj, I have changed code according to your comment, and below is the test results:

**sidecarInjectorWebhook.enableNamespacesByDefault is true:** 

```
$ istioctl install --set profile=demo -y --set values.sidecarInjectorWebhook.enableNamespacesByDefault=true
$ istioctl analyze -A
......
Info [IST0118] (Service calico-kube-controllers-metrics.calico-system) Port name metrics-port (port: 9094, targetPort: 9094) doesn't follow the naming convention of Istio port.
Info [IST0148] (Namespace bookinfo) neither injection label is set for namespace bookinfo, but Istio is installed with enableNamespacesByDefault enabled.
......
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.12/docs/reference/config/analysis for more information about causes and resolutions.
```

**Comment by irisdingbj** [comment]
Info [IST0148] (Namespace bookinfo) neither injection label is set for namespace bookinfo, but Istio is installed with enableNamespacesByDefault enabled     -->  
Info [IST0148] (Namespace bookinfo) is enabled for Istio injection,  as Istio is installed with `enableNamespacesByDefault` as `true`.

**Comment by zhlsunshine** [comment]
> Info [IST0148] (Namespace bookinfo) neither injection label is set for namespace bookinfo, but Istio is installed with enableNamespacesByDefault enabled -->
> Info [IST0148] (Namespace bookinfo) is enabled for Istio injection, as Istio is installed with `enableNamespacesByDefault` as `true`.

done!  Below is the result:

```
```
$ istioctl install --set profile=demo -y --set values.sidecarInjectorWebhook.enableNamespacesByDefault=true
$ istioctl analyze -A
......
Info [IST0118] (Service calico-kube-controllers-metrics.calico-system) Port name metrics-port (port: 9094, targetPort: 9094) doesn't follow the naming convention of Istio port.
Info [IST0148] (Namespace bookinfo) is enabled for Istio injection, as Istio is installed with enableNamespacesByDefault as true.
......
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.12/docs/reference/config/analysis for more information about causes and resolutions.
```
```

**Review by irisdingbj — APPROVED** [review]
Thanks!

**Pull request merged**